### PR TITLE
feat(partition): small multiples

### PR DESCRIPTION
**Pull request by monfera**
## Summary

Small multiples for partition charts

resize

image

- Additive small multiples (first level of small multiples; `index`): multiple `<Partition>` specs under `<Chart>` WIP
- Multiplicative small multiples (second level of small multiples; `innerIndex`): breakdown of a `<Partition>` spec along its first partitioning layer
- Both layers can independently specify horizontal, vertical or Z pattern (grid) (additive layer: not exposed yet)
- Eventually, 2D grid will be possible not just with Z pattern but also, allowing a category specification per dimension (additive) and allowing the breakdown along the second partitioning layer; not in scope for this PR
- Uses `getChartIdSelector`s in place of previous repeated `(state) => state.chartId`s (earlier unrelated feedback by Nick)
- Abstracts out some types, eg. `PanelPlacement`, `RelativeMargins`
- Tightens `containerBackgroundColor` in `getShapeViewModel` (no longer optional)
- BREAKING: redefines the inner/outer padding notation for small multiples from `[outer, inner]` to `{outer, inner}`

Warning, will likely do `push -f` while it's WIP

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [x] This was checked for cross-browser compatibility
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [X] Unit tests were updated or added to match the most common scenarios
- [x] Threading through small multiples and group by options into the spec
- [x] Fixing `margin`
- [X] Adding per-panel titles ~~and adding title info to the tooltip~~
- [X] Pie size related fixes (it still preserves the room for an extra layer)
- [ ] Adjust legend generation to the new (cartesian-compatible) type of small multiples generation via `<GroupBy>`
- [ ] Crop panel title with `...` if it doesn't fit horizontally
- [ ] Take font spec for panel title from the theme, use that for title text rendering, and preserve vertical area for title in function of that

### CODE REVIEW

File: src/chart_types/partition_chart/layout/viewmodel/viewmodel.ts [hunk 1]
```diff
@@ -396,8 +426,8 @@ export function shapeViewModel(
         });
   const maxLinkedLabelTextLength = config.linkLabel.maxTextLength;
   const linkLabelViewModels = linkTextLayout(
-    width,
-    height,
+    panelInnerWidth, // todo check why we're not passing inter-margin size
+    panelInnerHeight, // todo check why we're not passing inter-margin size
```
Comment (monfera) [inline src/chart_types/partition_chart/layout/viewmodel/viewmodel.ts:430]: now obsolete comment, will remove

**Review by markov00 — COMMENTED** [review]
There are two outstanding issues that I'm seeing here:

Screenshot 2021-03-23 at 11 53 47

With a vertical layout and 0 outer vertical padding, the first title is cutoff.
The legend instead mix up colors, in the example the the Mineral/fuel has a blue color, but is represented with the red one

File: src/chart_types/partition_chart/state/selectors/drilldown_active.ts [hunk 2]
```diff
@@ -20,9 +20,11 @@
...
 import { isLinear, isSimpleLinear } from '../../layout/viewmodel/viewmodel';
-import { getPartitionSpec } from './partition_spec';
+import { getPartitionSpecs } from './partition_spec';
 
 /** @internal */
-export const drilldownActive = createCachedSelector([getPartitionSpec], (spec) => {
-  return spec && isLinear(spec.config.partitionLayout) && isSimpleLinear(spec.config, spec.layers);
+export const drilldownActive = createCachedSelector([getPartitionSpecs], (specs) => {
+  return (
+    specs.length === 1 && isLinear(specs[0].config.partitionLayout) && isSimpleLinear(specs[0].config, specs[0].layers)
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/state/selectors/drilldown_active.ts:28]: Thanks for the catch, by now `isSimpleLinear`also starts with the quick `isLinear` test so it became redundant here. In this PR the addition is just the `[0]` index but good time to remove the 1st check

**Comment by monfera** [comment]
The vertical offsetting is improved to avoid cropping with a zero outer panel pad: cb12f05af0c3b0bb2008ea18ca183db2cc3329b4 and the Theme iteration (follow-up PR) will use the Theme font and padding specs

File: src/chart_types/partition_chart/state/selectors/drilldown_active.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/state/selectors/drilldown_active.ts]: done in 5f2e3bb53d3ecd986eca6f0ca43fa93fd8831c41

**Review by markov00 — APPROVED** [review]
LGTM, merging on green.
Followup PR:
- fix hierarchical legend

**Pull request merged**